orchestrator/search/retrieval/utils.py

```diff
@@ -0,0 +1,106 @@
+import json
+import re
+
+import structlog
+from sqlalchemy import and_
+from sqlalchemy_utils.types.ltree import Ltree
+
+from orchestrator.db.database import WrappedSession
+from orchestrator.db.models import AiSearchIndex
+from orchestrator.search.core.types import EntityType
+from orchestrator.search.indexing.registry import ENTITY_CONFIG_REGISTRY
+from orchestrator.search.schemas.parameters import BaseSearchParameters
+from orchestrator.search.schemas.results import SearchResult
+
+logger = structlog.get_logger(__name__)
+
+
+def generate_highlight_indices(text: str, term: str) -> list[tuple[int, int]]:
+    """Finds all occurrences of individual words from the term with word boundary matching case-insensitively."""
+    if not text or not term:
+        return []
+
+    all_matches = []
+    words = [w.strip() for w in term.split() if w.strip()]
+
+    for word in words:
+        word_boundary_pattern = rf"\b{re.escape(word)}\b"
+        matches = list(re.finditer(word_boundary_pattern, text, re.IGNORECASE))
+
+        if not matches:
+            substring_pattern = re.escape(word)
+            matches = list(re.finditer(substring_pattern, text, re.IGNORECASE))
```

Comment:
If a resulting text has both word _and_ substring matches, wouldn't we want to highlight the substring matches as well?